dj/api/nodes.py

```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
     List the available nodes.
     """
     return session.exec(select(Node)).all()
+
+
+@router.post("/nodes/", response_model=Union[SourceRead, NodeRead])
+def create_a_node(
+    data: Union[SourceCreate, NodeCreate],
+    session: Session = Depends(get_session),
+) -> Union[SourceRead, NodeRead]:
+    """
+    Create a node
+    """
+    dj_settings = get_settings()
+    # Check if the node already exists
+    query = select(Node).where(Node.name == data.name)
+    node = session.exec(query).one_or_none()
+    if node:
+        raise HTTPException(
+            status_code=HTTPStatus.CONFLICT,
+            detail=f"Node already exists: {data.name}",
+        )
+
+    if data.type == NodeType.SOURCE:
+        return create_source_node(session=session, data=data)
+
+    new_node = Node.from_orm(data)
+    new_nodes_parents = get_dependencies(new_node.query) if new_node.query else set()
+
+    # Check the columns for any dimension references and do a metadata lookup of the dimension node
+    for column in new_node.columns:
+        if dimension_target := column.get("dimension"):
+            dimension_target, dimension_column = dimension_target.rsplit(".", 1)
+            new_nodes_parents.add(column["dimension"])
+            query = (
+                select(Node)
+                .where(Node.name == dimension_target)
+                .where(Node.type == NodeType.DIMENSION)
+            )
+            dimension = session.exec(query).one_or_none()
+            if not dimension and dj_settings.fail_on_unknown_dimension_tag:
+                raise Exception(f"Invalid dimension: {dimension_target}")
+            column.dimension = dimension
+            column.dimension_column = dimension_column
+
+    # Lookup the parent nodes for the node
+    new_node.parents = session.exec(
+        select(Node).where(Node.name.in_(new_nodes_parents)),  # type: ignore  # pylint: disable=no-member
+    ).all()
+    _logger.info("Found parent nodes for %s: %s", new_node.name, new_node.parents)
+
+    # Infer the node's column names and column types by inspecting the parent nodes
+    try:
+        new_node.columns = infer_columns(new_node.query, parents=new_node.parents)
+    except Exception as e:  # pylint: disable=broad-except
+        error_message = f"Cannot infer columns for {new_node.name}: {str(e)}"
+        _logger.error(error_message)
+        if dj_settings.fail_on_infer_columns_error:
+            raise HTTPException(
+                status_code=HTTPStatus.PRECONDITION_FAILED,
+                detail=error_message,
+            ) from e
+
+    new_node.extra_validation()
+
+    session.add(new_node)
+    session.commit()
+    session.refresh(new_node)
+
+    return new_node  # type: ignore
+
+
+def create_source_node(session: Session, data: SourceCreate) -> Node:
+    """
+    Create a source node
+    """
+    dj_settings = get_settings()
+
+    config = {
```

Comment:
Should we add a marker for the client provided source nodes, say `mock: true`. This setting would prevent the DJ Watchdog from reporting on these source nodes as errors, we could still check them but not wake anyone up if a mock source node is not matching the underlying databases/tables?

Reply:
That's a good point. I'll need to think about this some more but I added it as an item to the weekly sync. Allowing CRUD through the API changes a few things that we'd have to consider, for example we will no longer have the signal of file changes to know when DJ should perform a validation analysis on the DAG.

We also have to think about how we handle situations where an upstream node is deleted or modified, making a once valid node now invalid (so if it has a `mock:false` attribute we'd need to know to switch it to `mock:true`). We could run the validation across the entire DAG each time a CRUD API request completes but that seems like the extreme solution and won't scale nicely. This is a tricky item with some nuances but a big one so it's worth all of us brainstorming a ton on it.

Reply:
I think ideally we want to run the validation across the DAG on every modification. Why do you think this wouldn't scale, @samredai? The validation can just track column lineage, so it should be fast and not require any database calls.

Reply:
That might just be an incorrect assumption on my part but I imagined it's more than just upstream column lineage, right? If we allow for draft/incomplete/mock nodes, then we'd have to check all of them every time a node is added to see if those nodes are now valid. We would even have to do this when a source node is added which has no upstream nodes. So I'm imagining what that would look like for a large DAG with many of these incomplete nodes.

Reply:
I've updated the PR with the idea from our discussion earlier to have a staging environment which has softer validation requirements for adding nodes. That punts the discussion in this thread to one of the next PRs when we add the logic to move a node from the staging environment to the production environment and have to perform the full validation. I like how much better this turned out, thanks @betodealmeida @agorajek @CircArgs!